---
sidebar_position: 8
description: "Stop infinite retries with Terminal Exceptions."
---

import Admonition from "@theme/Admonition";
import {CodeWithTabs} from '../../../src/components/code/code';

# Error Handling

Restate handles retries for failed invocations.
By default, Restate does infinite retries with an exponential backoff strategy.

For failures for which you do not want retries, but instead want the invocation to end and the error message
to be propagated back to the caller, you can throw a **terminal error**.

You can throw a terminal exception with an HTTP status code and a message anywhere in your handler, as follows:

<CodeWithTabs groupId={"sdk"}>
    ```java !!tabs Java
    CODE_LOAD::java/src/main/java/develop/ErrorHandling.java
    ```
    ```kotlin !!tabs Kotlin
    CODE_LOAD::kotlin/src/main/kotlin/develop/ErrorHandling.kt
    ```
</CodeWithTabs>

You can catch terminal exceptions. For example, you can catch the terminal exception that comes out of a [call to another service](/develop/java/service-communication#request-response-calls), and build your control flow around it.


<Admonition type="note" title="Error handling in Restate handlers">
    Have a look at our [Error Handling guide](/guides/error-handling) to learn more about how Restate handles transient errors, terminal errors, retries, and timeouts.
</Admonition>

<Admonition type="note" title="Sagas with Restate">
    When you throw a terminal error, you might need to undo the actions you did earlier in your handler to make sure that your system remains in a consistent state.
    Have a look at our [sagas guide](/guides/sagas) to learn more.
</Admonition>
